DataDog/datadog-agent
# Add config option to automatically detect common multi-line patterns when tailing logs.

### What does this PR do?

NOTE: @gh123man has taken over this feature. Please direct feedback/questions to @gh123man

This PR introduces a new feature for the logs agent to automatically detect common multi line patterns and combine multi-line logs (such as stack traces) into a single line when a pattern is detected. 

The approach used is to scan incoming logs against a list of known common patterns (regexs). If a pattern matches enough messages during the sample window (first 1000 logs or 30 seconds - whichever comes first) the single line handler will be replaced with a multiline handler with the selected pattern. 

When a new log source is detected - the logs are sent to Datadog with the single line handler while the agent tests the known patterns. If a pattern is selected all future logs will be processed with the multi line handler for that source even if the file rotates. However when the agent restarts, the detection will be run again. 

This feature is disabled by default for now under the config `logs_config.auto_multi_line_detection`. 
Additional patterns can be added under the `logs_config.auto_multi_line_extra_patterns` config. 
This feature can also be enabled/disabled per integration with the config `auto_multi_line_detection`. 
There are other options to tweak the behavior of this feature however they should be considered experimental and may change in future versions. 
 
### QA steps
## Setup
1. Create a working directory where your logs are going to be placed. 
2. Set your agent log level to debug
3. Create an integration config for your agent to pick up logs from this directory. Something like:
```yaml
logs:
  - type: file
    path: "/path/to/working/dir/*.log"
    service: "logs-qa"
    source: "logs-qa"
```
4. Create a script called `autorotate.sh` with the following content:
```bash
#!/bin/bash

i=0
while read x; do
    ((i=i+1))
    if ! ((i % $1)); then
        mv autorotate.log autorotate.log.0
    fi
    echo "$x" >> autorotate.log
done
```

## Scenarios 

1. start your agent
2. in your working directory, run `docker run -it bfloerschddog/java-excepton-logger | ./autorotate.sh 1000`
3. tail the agent logs and look for the following log line: 
```
 Pattern ^\d+-\d+-\d+ \d+:\d+:\d+(,\d+)? matched 965 lines with a ratio of 0.965000
```
4. after a few seconds you should see: 
```
 Auto multi line log detection enabled
 Found a previously detected pattern - using multi line handler
 ```
 5. let it run a little longer and this message should be repeated. This means that it is re-using the pattern correctly between file rotations 
 6. look at livetail for your host in app and confirm that the first logs submitted are not multi-line. But the later logs (after the first 1000) should be correctly combined. 

Repeat the above steps with the following containers:
`bfloerschddog/python-exception-logger` - looking for exceptions (like above)
`bfloerschddog/java-every-other-logger` - looking for every log line to be combine (instead of just exceptions)

## Settings and flags

1. test `auto_multi_line_detection` enable/disable globally
2. test `auto_multi_line_detection` enable/disable per integration 
3. test `auto_multi_line_extra_patterns`
 - add a new regex to the list in global config `\d+\.\d+\.\d+\.\d+` (to match IP addresses)
 - feed the agent logs that match this regex: `docker run -it --rm mingrammer/flog -l | ./autorotate.sh 1000`
    (note flog's default logs start with an IP address)
 - confirm the regex is used by looking for a log line (confirm it is indeed your new custom pattern): 
 ```
 Pattern ^\d+\.\d+\.\d+\.\d+ matched 1000 lines with a ratio of 1.00000
  ```
  Note: flog only produces single line logs so no need to look for multi-line combine logs. just testing that the user regex worked. 

 4. set `container_collect_all: true` and test that logs read from the docker socket also match a pattern. 
  - run `docker run -it bfloerschddog/java-every-other-logger`
  - look at agent logs for: `Pattern ^[A-Za-z_]+ \d+, \d+ \d+:\d+:\d+ (AM|PM) matched 500 lines with a ratio of 0.500000`
